src/common/proto_type.hpp

```diff
@@ -12,7 +12,7 @@ class ProtoType {
                    std::string,
                    int,
                    double,
-                   std::unordered_map<std::string, ProtoType*>,
+                   std::shared_ptr<std::unordered_map<std::string, std::shared_ptr<ProtoType>>>,
```

Comment:
is there a reason we aren't using `AttributeMap` here?

Reply:
I tried but I'm getting `use of undeclared identifier 'ProtoType'` and ` error: constructor cannot be redeclared`. might be a circular dependency?